File: FBSimulatorControl/Commands/FBSimulatorProcessSpawnCommands.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import "FBSimulatorProcessSpawnCommands.h"

#import <CoreSimulator/SimDevice.h>

#import <FBControlCore/FBControlCore.h>

#import "FBSimulator+Private.h"
#import "FBSimulator.h"
#import "FBSimulatorError.h"

@interface FBSimulatorProcessSpawnCommands ()

@property (nonatomic, weak, readonly) FBSimulator *simulator;

@end

@implementation FBSimulatorProcessSpawnCommands

#pragma mark Initializers

+ (instancetype)commandsWithTarget:(FBSimulator *)targets
{
  return [[self alloc] initWithSimulator:targets];
}

- (instancetype)initWithSimulator:(FBSimulator *)simulator
{
  self = [super init];
  if (!self) {
    return nil;
  }

  _simulator = simulator;

  return self;
}

#pragma mark FBSimulatorProcessSpawnCommands Implementation

- (FBFuture<FBProcess *> *)launchProcess:(FBProcessSpawnConfiguration *)configuration
{
  FBSimulator *simulator = self.simulator;

  return [[configuration.io
    attach]
    onQueue:simulator.workQueue fmap:^(FBProcessIOAttachment *attachment) {
      return [FBSimulatorProcessSpawnCommands
        launchProcessWithSimulator:simulator
        configuration:configuration
        attachment:attachment];
    }];
}

#pragma mark Public

+ (NSDictionary<NSString *, id> *)launchOptionsWithArguments:(NSArray<NSString *> *)arguments environment:(NSDictionary<NSString *, NSString *> *)environment waitForDebugger:(BOOL)waitForDebugger
{
  NSMutableDictionary<NSString *, id> *options = [NSMutableDictionary dictionary];
  options[@"arguments"] = arguments;
  options[@"environment"] = environment ? environment: @{@"__SOME_MAGIC__" : @"__IS_ALIVE__"};
  if (waitForDebugger) {
    options[@"wait_for_debugger"] = @1;
  }
  return options;
}

#pragma mark Private

+ (FBFuture<FBProcess *> *)launchProcessWithSimulator:(FBSimulator *)simulator configuration:(FBProcessSpawnConfiguration *)configuration attachment:(FBProcessIOAttachment *)attachment
{
  // Prepare captured futures
  id<FBControlCoreLogger> logger = simulator.logger;
  FBMutableFuture<NSNumber *> *launchFuture = [FBMutableFuture futureWithNameFormat:@"Launch of %@ on %@", configuration.launchPath, simulator.udid];
  FBMutableFuture<NSNumber *> *statLoc = [FBMutableFuture futureWithNameFormat:@"Process completion of %@ on %@", configuration.launchPath, simulator.udid];
  FBMutableFuture<NSNumber *> *exitCode = [FBMutableFuture futureWithNameFormat:@"Process exit of %@ on %@", configuration.launchPath, simulator.udid];
  FBMutableFuture<NSNumber *> *signal = [FBMutableFuture futureWithNameFormat:@"Process signal of %@ on %@", configuration.launchPath, simulator.udid];

  // Get the Options
  NSDictionary<NSString *, id> *options = [self
    simDeviceLaunchOptionsWithSimulator:simulator
    launchPath:configuration.launchPath
    arguments:configuration.arguments
    environment:configuration.environment
    waitForDebugger:NO
    stdOut:attachment.stdOut
    stdErr:attachment.stdErr
    mode:configuration.mode];

  // The Process launches and terminates asynchronously.
  [simulator.device
    spawnAsyncWithPath:configuration.launchPath
    options:options
    terminationQueue:simulator.workQueue
    terminationHandler:^(int stat_loc) {
      // Notify that we're done with the process to each of the futures.
      [FBProcessSpawnCommandHelpers
        resolveProcessFinishedWithStatLoc:stat_loc
        inTeardownOfIOAttachment:attachment
        statLocFuture:statLoc
        exitCodeFuture:exitCode
        signalFuture:signal
        processIdentifier:[launchFuture.result intValue]
        configuration:configuration
        queue:simulator.workQueue
        logger:logger];

      // Close any open file handles that we have.
      // This is important because otherwise any reader will stall forever.
      // The SimDevice APIs do not automatically close any file descriptor passed into them, so we need to do this on it's behalf.
      // This would not be an issue if using simctl directly, as the stdout/stderr of the simctl process would close when the simctl process terminates.
      // However, using the simctl approach, we don't get the pid of the spawned process, this is merely logged internally.
      // Failing to close this end of the file descriptor would lead to the write-end of any pipe to not be closed and therefore it would leak.
      
      [attachment.stdOut close];
      [attachment.stdErr close];
    }
    completionQueue:simulator.workQueue
    completionHandler:^(NSError *innerError, pid_t processIdentifier){
      if (innerError) {
        [launchFuture resolveWithError:innerError];
      } else {
        [launchFuture resolveWithResult:@(processIdentifier)];
      }
  }];

  // Map to the FBProcess implementation.
  return [launchFuture
    onQueue:simulator.workQueue map:^(NSNumber *processIdentifierNumber) {
      // Wrap in the container object
      pid_t processIdentifier = processIdentifierNumber.intValue;
      return [[FBProcess alloc] initWithProcessIdentifier:processIdentifier statLoc:statLoc exitCode:exitCode signal:signal configuration:configuration queue:simulator.workQueue];
    }];
}

+ (NSDictionary<NSString *, id> *)simDeviceLaunchOptionsWithSimulator:(FBSimulator *)simulator launchPath:(NSString *)launchPath arguments:(NSArray<NSString *> *)arguments environment:(NSDictionary<NSString *, NSString *> *)environment waitForDebugger:(BOOL)waitForDebugger stdOut:(nullable FBProcessStreamAttachment *)stdOut stdErr:(nullable FBProcessStreamAttachment *)stdErr mode:(FBProcessSpawnMode)mode
{
  // argv[0] should be launch path of the process. SimDevice does not do this automatically, so we need to add it.
  arguments = [@[launchPath] arrayByAddingObjectsFromArray:arguments];
  NSMutableDictionary<NSString *, id> *options = [[self launchOptionsWithArguments:arguments environment:environment waitForDebugger:waitForDebugger] mutableCopy];
  if (stdOut){
    options[@"stdout"] = @(stdOut.fileDescriptor);
  }
  if (stdErr) {
    options[@"stderr"] = @(stdErr.fileDescriptor);
  }
  options[@"standalone"] = @([self shouldLaunchStandaloneOnSimulator:simulator mode:mode]);
  return [options copy];
}

+ (BOOL)shouldLaunchStandaloneOnSimulator:(FBSimulator *)simulator mode:(FBProcessSpawnMode)mode
{
  // Standalone means "launch directly, not via launchd"
  switch (mode) {
    case FBProcessSpawnModeLaunchd:
      return NO;
    case FBProcessSpawnModePosixSpawn:
      return YES;
    default:
      // Default behaviour is to use launchd if booted, otherwise use standalone.
      return simulator.state != FBiOSTargetStateBooted;
  }
}

@end

File: FBSimulatorControl/Commands/FBSimulatorProcessSpawnCommands.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

@_implementationOnly import CoreSimulator
@preconcurrency import FBControlCore
import Foundation

public final class FBSimulatorProcessSpawnCommands: NSObject, FBiOSTargetCommand {

  // MARK: - Properties

  private weak var simulator: FBSimulator?

  // MARK: - Initializers

  public class func commands(with target: any FBiOSTarget) -> FBSimulatorProcessSpawnCommands {
    return FBSimulatorProcessSpawnCommands(simulator: target as! FBSimulator)
  }

  private init(simulator: FBSimulator) {
    self.simulator = simulator
    super.init()
  }

  // MARK: - Launch Options

  class func launchOptions(withArguments arguments: [String], environment: [String: String], waitForDebugger: Bool) -> [String: Any] {
    var options: [String: Any] = [:]
    options["arguments"] = arguments
    options["environment"] = environment
    if waitForDebugger {
      options["wait_for_debugger"] = NSNumber(value: 1)
    }
    return options
  }

  // MARK: - Private

  fileprivate func launchProcess(_ configuration: FBProcessSpawnConfiguration) async throws -> FBSubprocess<AnyObject, AnyObject, AnyObject> {
    guard let simulator else {
      throw FBSimulatorError.describe("Simulator deallocated").build()
    }
    let attachment = try await bridgeFBFuture(configuration.io.attach())
    return try await FBSimulatorProcessSpawnCommands.launchProcess(
      withSimulator: simulator,
      configuration: configuration,
      attachment: attachment
    )
  }

  private class func launchProcess(withSimulator simulator: FBSimulator, configuration: FBProcessSpawnConfiguration, attachment: FBProcessIOAttachment) async throws -> FBSubprocess<AnyObject, AnyObject, AnyObject> {
    let logger = simulator.logger
    let statLoc = FBMutableFuture<NSNumber>(name: "Process completion of \(configuration.launchPath) on \(simulator.udid)")
    let exitCode = FBMutableFuture<NSNumber>(name: "Process exit of \(configuration.launchPath) on \(simulator.udid)")
    let signal = FBMutableFuture<NSNumber>(name: "Process signal of \(configuration.launchPath) on \(simulator.udid)")

    let options = simDeviceLaunchOptions(
      withSimulator: simulator,
      launchPath: configuration.launchPath,
      arguments: configuration.arguments,
      environment: configuration.environment,
      waitForDebugger: false,
      stdOut: attachment.stdOut,
      stdErr: attachment.stdErr,
      mode: configuration.mode
    )

    // PID is needed by the termination handler for logging; we populate the
    // holder once `spawnAsync` returns it. The terminationHandler will only be
    // invoked after the process exits, which strictly follows that return.
    let pidHolder = PIDHolder()
    let processIdentifier = try await simulator.device.spawnAsync(
      withPath: configuration.launchPath,
      options: options,
      terminationQueue: simulator.workQueue,
      terminationHandler: { (statLocValue: Int32) in
        FBProcessSpawnCommandHelpers.resolveProcessFinished(
          withStatLoc: statLocValue,
          inTeardownOfIOAttachment: attachment,
          statLocFuture: statLoc,
          exitCodeFuture: exitCode,
          signalFuture: signal,
          processIdentifier: pidHolder.value,
          configuration: configuration,
          queue: simulator.workQueue,
          logger: logger
        )
        attachment.stdOut?.close()
        attachment.stdErr?.close()
      },
      completionQueue: simulator.workQueue
    )
    pidHolder.value = processIdentifier

    return FBSubprocess<AnyObject, AnyObject, AnyObject>(
      processIdentifier: processIdentifier,
      statLoc: convertFBMutableFuture(statLoc),
      exitCode: convertFBMutableFuture(exitCode),
      signal: convertFBMutableFuture(signal),
      configuration: configuration,
      queue: simulator.workQueue
    )
  }

  /// Lets the termination handler reach the PID once it's known, since
  /// `spawnAsync` produces the PID after the handler is registered.
  private final class PIDHolder: @unchecked Sendable {
    var value: Int32 = 0
  }

  // Internal (not private) so the option dictionary can be characterized by unit tests; see FBSimulatorProcessSpawnCommandsTests.
  class func simDeviceLaunchOptions(withSimulator simulator: FBSimulator, launchPath: String, arguments: [String], environment: [String: String], waitForDebugger: Bool, stdOut: FBProcessStreamAttachment?, stdErr: FBProcessStreamAttachment?, mode: FBProcessSpawnMode) -> [String: Any] {
    // argv[0] should be launch path of the process. SimDevice does not do this automatically, so we need to add it.
    let fullArguments = [launchPath] + arguments
    var options = launchOptions(withArguments: fullArguments, environment: environment, waitForDebugger: waitForDebugger)
    if let stdOut {
      options["stdout"] = NSNumber(value: stdOut.fileDescriptor)
    }
    if let stdErr {
      options["stderr"] = NSNumber(value: stdErr.fileDescriptor)
    }
    options["standalone"] = NSNumber(value: shouldLaunchStandalone(onSimulator: simulator, mode: mode))
    return options
  }

  // Internal (not private) for unit-test characterization of the launchd-vs-standalone decision.
  class func shouldLaunchStandalone(onSimulator simulator: FBSimulator, mode: FBProcessSpawnMode) -> Bool {
    switch mode {
    case .launchd:
      return false
    case .posixSpawn:
      return true
    default:
      // Default behaviour is to use launchd if booted, otherwise use standalone.
      return simulator.state != .booted
    }
  }
}

// MARK: - FBSimulator+ProcessSpawnCommands

extension FBSimulator: ProcessSpawnCommands {

  public func launchProcess(
    _ configuration: FBProcessSpawnConfiguration
  ) async throws -> FBSubprocess<AnyObject, AnyObject, AnyObject> {
    try await processSpawnCommands().launchProcess(configuration)
  }
}
